api/src/controllers/africas-talking.js

```diff
@@ -0,0 +1,98 @@
+const serverUtils = require('../server-utils');
+
+// Map from Africa's Talking status codes to ours
+// https://build.at-labs.io/docs/sms%2Fnotifications
+// https://github.com/medic/medic-docs/blob/master/user/message-states.md
+const STATUS_CODES = {
+  // Sent: The message has successfully been sent by our network.
+  Sent: 'forwarded-by-gateway',
+
+  // Submitted: The message has successfully been submitted to the MSP (Mobile Service Provider).
+  Submitted: 'sent',
+
+  // Buffered: The message has been queued by the MSP.
+  Buffered: 'sent',
+
+  // Rejected: The message has been rejected by the MSP. This is a final status.
+  Rejected: 'failed',
+
+  // Success: The message has successfully been delivered to the receiver's handset. This is a final status.
+  Success: 'delivered',
+
+  // Failed: The message could not be delivered to the receiver's handset. This is a final status.
+  Failed: 'failed'
+};
+
+// Stolen from: https://github.com/expressjs/express/issues/2518#issuecomment-455775771
+const isLocal = req => req.connection.localAddress === req.connection.remoteAddress;
+
+const validateSource = req => {
+  // allow "http" requests only from localhost
+  if (isLocal(req)) {
+    return;
+  }
+  if (req.protocol !== 'https') {
+    // this should be blocked by our stack but adding here to double
+    // check and prevent IP spoofing
+    logger.warn(`Attempt to access Africa's Talking endpoint with protocol: "${req.protocol}", ${req.ip}`);
+    return { code: 403, message: 'https is required' };
+  }
+  const settings = config.get('sms');
+  const ips = settings &&
+              settings.africas_talking &&
+              settings.africas_talking.allowed_ips;
+  if (!ips || !Array.isArray(ips) || !ips.includes(req.ip)) {
+    logger.warn(`Attempt to access Africa's Talking endpoint by IP address not on allowed list: "${req.ip}"`);
+    return { code: 403, message: 'Unknown IP address' };
+  }
+};
+
+const checkAuth = (req, res) => {
+  const err = validateSource(req);
+  if (err) {
+    serverUtils.error(err, req, res);
+    return false;
+  }
+  return true;
+};
+
+module.exports = {
+  incomingMessages: (req, res) => {
+    if (!checkAuth(req, res)) {
+      return;
+    }
+    if (!req.body) {
+      return;
+    }
+    const message = {
+      id: req.body.id,
+      from: req.body.from,
+      content: req.body.text
+    };
+    return messaging.processIncomingMessages([ message ])
+      .then(() => res.json({ success: true }))
+      .catch(err => serverUtils.error(err, req, res));
+  },
+  deliveryReports: (req, res) => {
+    if (!checkAuth(req, res)) {
+      return;
+    }
+    const state = STATUS_CODES[req.body.status];
+    if (!state) {
+      return serverUtils.error({
+        code: 400,
+        message: `Unknown status code: "${req.body.status}", gateway message reference: "${req.body.id}"`
+      }, req, res);
+    }
+    return messaging.updateMessageTaskStates([{
```

Comment:
Is it the desired behavior to respond with `{ success: true }` if the message is not found?

Reply:
I'm not sure what the desired behaviour is here.

My logic was - AT has done all they can. We don't want them to retry over and over again so we should tell them we received the update even if it didn't match anything we have.

It would probably be clearer if I returned `{ success: false }` or something else which they ignore but anyone else debugging the response would understand.

 The Africa's Talking API docs didn't mention retries or expected responses but I'll check with them to see if they check for anything, then come up with a plan from there.

Reply:
Yea, I had a look at the docs as well and noticed the same lack of information. This was just a question of consistency or correctness of one of _our_ API endpoints.

Reply:
I'm now returning the number of updated docs.